hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java

```diff
@@ -132,252 +141,267 @@ public Traverser(List<RepeatEdgeStep> steps,
-            traverseIds(this.sources.keySet().iterator(), vid -> {
-                if (this.reachLimit()) {
-                    return;
+        private void processOneForBackward(Id source, Id target) {
+            for (Node n : this.targets.get(source)) {
+                // If have loop, skip target
+                if (!this.withRing && n.contains(target)) {
+                    continue;
                 }
-                Iterator<Edge> edges = edgesOfVertex(vid, finalCurrentStep);
-                while (edges.hasNext()) {
-                    HugeEdge edge = (HugeEdge) edges.next();
-                    Id target = edge.id().otherVertexId();
-
-                    for (Node n : this.sources.get(vid)) {
-                        // If have loop, skip target
-                        if (!this.withRing && n.contains(target)) {
-                            continue;
-                        }
 
-                        // If cross point exists, path found, concat them
-                        if (this.lastSuperStep() &&
-                            this.targets.containsKey(target)) {
-                            for (Node node : this.targets.get(target)) {
-                                List<Id> path = n.joinPath(node);
-                                if (!path.isEmpty()) {
-                                    this.paths.add(new Path(target, path));
-                                    if (this.reachLimit()) {
-                                        return;
-                                    }
-                                }
+                // If cross point exists, path found, concat them
+                if (this.lastSuperStep() &&
+                    this.sourcesAll.containsKey(target)) {
+                    for (Node node : this.sourcesAll.get(target)) {
+                        List<Id> path = joinPath(n, node, this.withRing);
+                        if (!path.isEmpty()) {
+                            Path newPath = new Path(target, path);
+                            newPath.reverse();
+                            this.paths.add(newPath);
+                            if (this.reachLimit()) {
+                                return;
                             }
                         }
-
-                        // Add node to next start-nodes
-                        newVertices.add(target, new Node(target, n));
                     }
                 }
-            });
 
-            for (Map.Entry<Id, List<Node>> entry : newVertices.entrySet()) {
-                this.sourcesAll.addAll(entry.getKey(), entry.getValue());
+                // Add node to next start-nodes
+                this.addNodeToNewVertices(target, new Node(target, n));
+            }
+        }
+
+        private void reInitAllIfNeeded(boolean forward) {
+            if (forward) {
+                // Re-init source all if last forward finished one super step
+                // and not last super step
```

Comment:
not finished last super step? or current step is not the last super step?
and use /*